test/LiquidityPool.t.sol

```diff
@@ -68,6 +71,80 @@ contract LiquidityPoolTest is Test {
         mockXcmRouter.file("gateway", address(gateway));
         evmInvestmentManager.rely(address(gateway));
         escrow.rely(address(gateway));
+
+        self = address(this);
+    }
+
+    function testTransferFrom(
+        uint64 poolId,
+        uint8 decimals,
+        string memory tokenName,
+        string memory tokenSymbol,
+        bytes16 trancheId,
+        uint128 price,
+        uint128 currencyId,
+        uint256 amount,
+        uint64 validUntil
+    ) public {
+        vm.assume(currencyId > 0);
+        vm.assume(amount < MAX_UINT128);
+        vm.assume(amount > 100);
+        vm.assume(validUntil >= block.timestamp);
+        price = 1;
+        Investor investor = new Investor();
+
+        address lPool_ = deployLiquidityPool(poolId, decimals, tokenName, tokenSymbol, trancheId, price, currencyId);
+        investorDeposit(address(investor), lPool_, poolId, trancheId, amount, validUntil); // deposit funds first
+        LiquidityPool lPool = LiquidityPool(lPool_);
+        homePools.updateMember(poolId, trancheId, self, validUntil); // put self on memberlist to be able to receive tranche tokens
+
+        // investor fails to transfer tranche tokens no approval on LPool
+        vm.expectRevert(bytes("ERC20/insufficient-allowance")); // Todo -> discuss to add auto-approval on transfer
+        investor.transferFrom(lPool_, address(investor), self, amount / 4);
+
+        // investor can transfer tranche tokens
+        investor.approve(lPool_, lPool_, type(uint256).max);
+        investor.transferFrom(lPool_, address(investor), self, amount / 4);
+
+        // Random user (self) can not transfer tokens on behalf of user
+        vm.expectRevert(bytes("LiquidityPool/no-token-allowance"));
+        lPool.transferFrom(address(investor), self, amount / 4);
+
+        // Random user (self) can transfer tokens on behalf of user after approval granted
+        investor.approve(lPool_, self, amount / 4);
+        lPool.transferFrom(address(investor), self, amount / 4);
```

Comment:
Personally, I think adding some balance assertions here would be nicer than relying on reverts.

Reply:
yeah, agree. In this test it is about permission setup though. Other tests using this functions have balance checks